feat: Include accessibility scans in ci

LilaKelland opened the pull request:
- [x] Add deque axe-linter for development for dev container
- [x] add accessibility module  (note needs refactoring and cleaning up)
- [x] allow for whitelisting axe-ids
- [x] allow for blacklisting urls (bot.txt) (right now just does it with config  - change to robot.txt, and also option to just take in a set list of urls to scan rather than crawling (haven't done this with bot.txt - should we do this as well?)
~~- [ ] Add tests~~ moving out to next PR
- [x] Integrate into cloud build ci
~~- [ ] possibly use Pa11yci as an alternative (with axe)~~ leaving aside for now  - the benefit to using this would be that it's simpler, but the check itself is not complicated with axe, and axe-core library is more comprehensive. 
- [x] save file into bucket - #TODO add to acm-core to be able to push to this project. 
~~- [ ] reflect results back on the Readme? (later?)~~ Leaving this off for now - pushing everything to a bucket and will leave the dashboard/reporting until later. 


Remaining known issues:

1. Writing to the as a non-root user in process-axe-report.js causes errors: 
```
const filename = `./axe-results/ci_axe_results_${timestamp}.json`;
fs.writeFileSync(filename, JSON.stringify(result, null, 2), 'utf8');
```
This was fixed by adding the prepare-results-directory step in ui/cloudbuild.yaml  - I feel like there is a better way to do this? 

2. Are the env variables cloudbuild store in Cloud secrets? I imagine the bucket name should be and ENV variable. 

3. Need to point the bucket in ui/cloudbuild.yaml to one in the safe-inputs GCP project (to prevent permission errors). 

4. Swap out pause in Cloudbuild for health check. (The more efficient image for axe-testing finishes before the docker compose environment has completed spinning up.)

(Added Issue https://github.com/HC-IPPM/safe-inputs/issues/782)

Stephen-ONeil commented on axe-testing/whitelist-config.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "blacklistUrls": [
+    "http://127.0.0.1:8080/termsAndConditions",
+    "http://127.0.0.1:8080/signin?post_auth_redirect=%2F&message=SessionRequired"
+  ],
```
I'd recommend blacklisting path patterns, rather than URLs. The base might be dependent on env (port might change), and there will be paths with parameterized portions which we might want to blacklist (e.g. the `post_auth_redirect` query parameter won't always be empty, or we might want to blacklist a route like `manage-collection/${collection id}/edit-column/${column id}`; could do both of these with regex patterns).

Stephen-ONeil commented on axe-testing/src/crawl-page.js:
```diff
@@ -0,0 +1,68 @@
+    results,
+  });
+
+  // Get all links on the current page
+  const links = await page.$$eval('a', (anchors) =>
+    anchors.map((anchor) => anchor.href),
+  );
+
+  // Crawl through each link found on the current page
+  for (const link of links) {
+    // Avoid revisiting the same pages or external URLs
+    if (!visitedPages.has(link) && link.startsWith(HOMEPAGE_URL)) {
+      const newPage = await browser.newPage();
+      await newPage.goto(link, { waitUntil: 'networkidle2' });
+      await crawlPage(
+        newPage,
+        browser,
+        HOMEPAGE_URL,
+        visitedPages,
+        allResults,
+        blacklistUrls,
+      ); // Recursively crawl new pages
+      await newPage.close(); // Close new page after crawling
+    }
+  }
```
This is good! It doesn't cover everything as you probably know (which is very hard to do with crawling). I'll recap some of what's not covered here, but this is just for reference and future refinement, not for trying to address in _this_ PR.

1) some routes only exist if the right data is present (e.g. links to collection management or upload routes from the home page only exist if the logged in user has collections to manage/upload to)
    - could be addressed by seeding the database with specially crafted data before performing the accessibility scan. It'll be a small maintenance burden, but it's that or try and make a crawler that can correctly fill out and submit valid forms, and then re-crawl the site for dynamic changes/new forms until everything has been reached... which is going to be slower and a nightmare to build, haha.
2) navigation can also occur via button clicks (e.g. when the server response includes a redirect, which we won't generally do, or when client side JS navigates the client based on server response, such as moving the user the the collection management route for a new collection after submitting a collection creation form)
    - off the top of my head, and ideally it stays this way, any non-link navigation leads to a page that is also reachable by link, if the data for it exists. Might be able to ignore this problem :crossed_fingers: 
3) elements that render/respond to user input won't be interacted with by a shallow scan on each route (e.g. owner and uploader email input fields don't exist until the user clicks the button to add new emails)
    - this is also likely not worth trying to fully automate as part of our end-to-end scans, that's too hard a problem. Likewise, maintaining puppeteer scripts to exercise dynamic elements on each page would be a big maintenance burden. I'd prefer to leave interactive elements out of scope in this end-to-end crawler, and instead figure out a unit test (component level) equivalent for more dynamic accessibility checks.

Stephen-ONeil replied:
Item 1 is the only one I'd prioritize for follow up. Happy to discuss it any time you want!

Stephen-ONeil reviewed (changes requested):
Small things mostly, and some comments on larger problems that are out-of-scope for this PR but could use some thought/consideration for follow up :+1:

Stephen-ONeil commented:
Also free to discuss the known issue bullets any time!

LilaKelland commented on axe-testing/eslint.config.js:
```diff
@@ -0,0 +1,7 @@
+import globals from 'globals';
+import pluginJs from '@eslint/js';
+
+export default [
+  { languageOptions: { globals: globals.browser } },
+  pluginJs.configs.recommended,
+];
```
Sounds good!  I tried to moved eslint back to the <9 version and copied the .eslintrc from the api (and installed the same eslint dev dependencies, but <9 is deprecated so followed the instructions to migrate (so now it has the eslint.config.mjs again that extends from the root (but left the old .eslintrc, as well. )  It doesn't work as there are now version conflicts with the eslint at the root... will see if the cloudbuild linting still works.  Should I migrate everything over, or leave this for now, and have migration as a separate activity?

Stephen-ONeil replied:
Ah, hm, leave for now in that case. Upgrading everything to eslint v9 has been held up for a while, see the comments in #229.

Although now that I'm looking at that PR, I think the last plugin I was waiting on/expecting to get v9 compatibility has updated since I last checked. Eitherway, it should be its own PR.

LilaKelland commented on axe-testing/package.json:
```diff
@@ -0,0 +1,28 @@
+    "test": "NODE_OPTIONS=--experimental-vm-modules jest --coverage",
+    "lint": "eslint src",
+    "dbg": "node --experimental-vm-modules --inspect-brk node_modules/.bin/jest --runInBand --no-cache",
+    "start": "node --experimental-vm-modules index.js"
```
added 
```
    "prestart:docker": "mkdir -p axe-results && sudo chmod 777 axe-results && docker build -t axe .",
    "start:docker": "docker run --rm --network host -v ${PWD}/axe-results:/app/axe-results axe"
```

LilaKelland commented:
Thanks Stephen!  Updated to reflect feedback but didn't touch the seeding the DB to trigger additional links to be found by the crawler yet.   Added it (and other items) here: https://github.com/HC-IPPM/safe-inputs/issues/782 .

Stephen-ONeil reviewed (changes requested):
Looks good! You'll need to merge in some changes on main, and you'll want to incorporate some of them I think.

I deleted the root package.json's `postinstall`, and added an explicit `ci:all` script for installing all sub dir dependencies. You'll need to move the axe-testing install there.

I've also added a script, `/wait_for_gql_service.cjs` that will retry (up to a timeout limit) to connect to a local GQL service, which should be more reliable than the `sleep 50` in the `ui/cloudbuild.yaml` here (should be... seems to be reaching the time out quite often right now, making our root ci flaky).

LilaKelland commented:
Good catch!  Removed .eslintrc.  Also tried to add the bucket to ui/cloudbuild as the acm-core github action bug was resolved this morning, and ran it to ensure it still works.  Running locally works still, but now getting ui-codegen / node-modules permission errors when running in ui/cloubuild when docker composing up.  Have gone down a rabbithole trying to fix, changing paths and permissions (haven't committed/pushed as it's still broken but in a different, probably not better, way) - wondering is there an eas(ier) fix for this?

Stephen-ONeil commented:
> Good catch! Removed .eslintrc. Also tried to add the bucket to ui/cloudbuild as the acm-core github action bug was resolved this morning, and ran it to ensure it still works. Running locally works still, but now getting ui-codegen / node-modules permission errors when running in ui/cloubuild when docker composing up. Have gone down a rabbithole trying to fix, changing paths and permissions (haven't committed/pushed as it's still broken but in a different, probably not better, way) - wondering is there an eas(ier) fix for this?

Hm, the error message is unclear to me too, mind if I hop on this branch and push some commits to try and figure it out?

Stephen-ONeil commented:
> Absolutely - thanks Stephen!

Done! There were two issues tripping things up:
1) the axe-testing package-lock had a version of `@microsoft/eslint-plugin-sdl` that wasn't compatible with eslint v8. That's my fault ultimately for introducing the pattern of pulling the package from GitHub without specifying a version, I vaguely think it wasn't on npm when I first added it, but could be misremembering. It is now, so I pinned it to version 0.2.2 explicitly.
2) seems the docker compose `ui-gql-codegen` service has read/write issues when running in cloudbuild (maybe because it tries to run as root?). Didn't notice before because I was scaling that step to 0 where I was running the dev compose env in cloud build as it isn't necessary, so that's the easy fix here too.

I also reorganized the cloud build steps, put all the setup tasks in to one step, mostly because it made it easier to use the `wait_for_gql_service.cjs` script (now a generic `wait_for_service.cjs` that can also wait for the ui server) in place of the 50 second timeout :+1:.

LilaKelland commented:
Thanks for the clean up and fixes!! 

I'd added in the save to bucket step (but then realized cloudbuild probably needs permissions to save to the bucket in safe inputs so commented it back out again).  

In the process, I tried running the docker compose dev environment with npm run dev but have been getting this error.
```
​Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "../node-dev-entrypoint.once.sh": stat ../node-dev-entrypoint.once.sh: no such file or directory: unknown
```

Running in cloudbuild works great, so can't figuring why there would be a path difference between cloudbuild and local?

Stephen-ONeil commented:
> In the process, I tried running the docker compose dev environment with npm run dev but have been getting this error.
> 
> ```
> ​Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "../node-dev-entrypoint.once.sh": stat ../node-dev-entrypoint.once.sh: no such file or directory: unknown
> ```

I'm guessing its the same/a related issue with the `ui-gql-codegen` service in the compose environment. Try commenting it out and running again. Dang, if it is then I guess my machine is just a special case where it works haha. I'll have to figure out a different way to set it up that works more generally...

LilaKelland commented:
Strange... unfortunately it still generates the same error with code-gen commented out in docker-compose.dev.yaml (in the db-1 service).

Stephen-ONeil commented:
> Strange... unfortunately it still generates the same error with code-gen commented out in docker-compose.dev.yaml (in the db-1 service).

:thinking: hm, the db service shouldn't be trying to use `node-dev-entrypoint.once.sh`.

Does the it work if you check out main? If it's not specific to this branch, we could move along for now and sort it out later today.

Stephen-ONeil commented:
> Unfortunately fails there too

That's ok, just means it's not a blocker on merging this branch. Let me know when you're ready for one last code review, I think it's ready to merge though!

Could you schedule a meeting this evening (after 3) or tomorrow for us to try and work out the docker compose problems you're seeing? Thanks!

LilaKelland commented:
Thanks! Ready to review.  I was going to remove the branch_name and commit_sha from cloudbuild submit instructions in axe-testing README (as save to gcs bucket where these are used is commented back out in ui/cloudbuild.yaml), decided to leave it for now as it doesn't break when run.  Unless it makes more sense to remove... Happy to change names/ naming conventions.

Stephen-ONeil reviewed (approved):
:+1: :tada:
